cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -302,11 +325,53 @@ private predicate fieldReadStep(Node node1, FieldContent f, Node node2) {
   )
 }
 
+predicate suppressArrayRead(Node node1, ArrayContent a, Node node2) {
+  a = TArrayContent() and
+  exists(BufferMayWriteSideEffectInstruction write, ChiInstruction chi |
+    node1.asInstruction() = write and
+    node2.asInstruction() = chi and
+    chi.getPartial() = write and
+    getWrittenField(write, _, _)
+  )
+}
+
+private class ArrayToPointerConvertInstruction extends ConvertInstruction {
+  ArrayToPointerConvertInstruction() {
+    this.getUnary().getResultType() instanceof ArrayType and
+    this.getResultType() instanceof PointerType
+  }
+}
+
 private predicate arrayReadStep(Node node1, ArrayContent a, Node node2) {
   a = TArrayContent() and
-  exists(LoadInstruction load |
-    node1.asInstruction() = load.getSourceValueOperand().getAnyDef() and
-    load = node2.asInstruction()
+  (
+    // In cases such as:
+    // ```cpp
+    // void f(int* pa) {
+    //   *pa = source();
+    // }
+    // ...
+    // int x;
+    // f(&x);
+    // use(x);
+    // ```
+    // the load on `x` in `use(x)` will exactly overlap with its definition (in this case the definition
+    // is a `BufferMayWriteSideEffect`).
+    exists(LoadInstruction load |
+      node1.asInstruction() = load.getSourceValue() and
+      load = node2.asInstruction()
+    )
+    or
+    // Explicit dereferences such as `*p` or `p[i]` where `p` is a pointer or array.
+    exists(LoadInstruction load |
+      node1.asInstruction() = load.getSourceValueOperand().getAnyDef() and
+      load = node2.asInstruction() and
+      (
+        load.getSourceAddress() instanceof LoadInstruction or
+        load.getSourceAddress() instanceof ArrayToPointerConvertInstruction or
+        load.getSourceAddress() instanceof PointerAddInstruction
+      )
```

Comment:
This list does not seem exhaustive. What are you really trying to specify? Is it simpler to write it in a negative way instead, like `not load.getSourceAddress() instanceof VariableAddressInstruction`?

Reply:
I think the list should be exhaustive? An array read is either of the form `*x` or `x[i]`, where `x` is either a pointer or an array that will undergo array-to-pointer conversion:
- `load.getSourceAddress() instanceof LoadInstruction` convers `*x` where `x` is a pointer
- `load.getSourceAddress() instanceof ArrayToPointerConvertInstruction` covers `*x` where `x` is an array
- `load.getSourceAddress() instanceof PointerAddInstruction` covers `x[i]`

Reply:
(I am now convinced that you were right)